tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp

```diff
@@ -104,23 +104,30 @@ void cq_reserve_back(uint32_t data_size_B) {
 }
 
 FORCE_INLINE
-void cq_push_back(const uint32_t command_issue_region_size, uint32_t push_size_B) {
+void notify_host_of_cq_completion_write_pointer() {
+    constexpr static uint64_t pcie_address = (uint64_t(NOC_XY_ENCODING(PCIE_NOC_X, PCIE_NOC_Y)) << 32) | HOST_CQ_COMPLETION_WRITE_PTR;  // For now, we are writing to host hugepages at offset
+    uint32_t completion_wr_ptr_and_toggle = cq_write_interface.completion_fifo_wr_ptr | (cq_write_interface.completion_fifo_wr_toggle << 31);
+    volatile tt_l1_ptr uint32_t* completion_wr_ptr_addr = get_cq_completion_write_ptr();
+    completion_wr_ptr_addr[0] = completion_wr_ptr_and_toggle;
+    noc_async_write(CQ_COMPLETION_WRITE_PTR, pcie_address, 4);
+    noc_async_write_barrier();
+}
+
+FORCE_INLINE
+void cq_push_back(const uint32_t command_issue_region_size, uint32_t push_size_B, uint32_t min_space_required_B) {
     uint32_t push_size_16B = align(push_size_B, 32) >> 4;
+    uint32_t min_space_required_16B = align(min_space_required_B, 32) >> 4;
```

Comment:
Im not sure if I should do this but there can be a case where we split our buffer read into multiple enqueue_read_buffer commands but during that process we could update our write ptr such that there isn't enough space for one page. In that case we can wrap the completion queue earlier than the next host iteration of computing how much available space there is by seeing if updated wr ptr would leave less than `min_space_required_B` in the completion queue. The completion queue read ptr is updated in same manner

Issue with this is that if the current read is the last read for the buffer and the next buffer to read has smaller page size than current, then we may have unnecessarily wrapped

If I remove this logic then host will just see there is not enough space and issue a wrap to completion (might be more straightforward to follow). Let me know your thoughts @DrJessop

Reply:
I'm not following. If we issue multiple enqueue read buffer commands, why can't we just have a loop that has the following:

```c++
while (not done) {
 if (wrap_condition)
  wrap();
 EnqueueReadBuffer command(...); 
 command.process()
}
```

Reply:
yes will update logic